### PR TITLE
Balancer with stake + Boost/Compensate abstraction

### PR DESCRIPTION
**Pull request by Shelvak**

### CODE REVIEW

File: contracts/ControllerBalancerV2Strat.sol [hunk 1]
```diff
@@ -180,8 +165,14 @@ contract ControllerBalancerV2Strat is ControllerStratAbs {
         uint[] memory amounts = new uint[](tokens.length);
 
         uint _balance = wantBalance();
-        uint bpt_balance = balanceOfPool();
+
+        //Unstake
+        uint staked_balance = balanceOfPool();
+        IBalancerGauge(GAUGE).withdraw(staked_balance);
+        require(balanceOfVaultPool() >= staked_balance, "Gauge gave less than expected");
```
Comment (batudo) [inline contracts/ControllerBalancerV2Strat.sol:172]:
```suggestion
        uint stakedBalance = balanceOfPool();
        IBalancerGauge(GAUGE).withdraw(stakedBalance);
        require(balanceOfVaultPool() >= stakedBalance, "Gauge gave less than expected");
```

File: contracts/ControllerBalancerV2Strat.sol [hunk 2]
```diff
@@ -180,8 +165,14 @@ contract ControllerBalancerV2Strat is ControllerStratAbs {
...
+
         uint index = 0;
+        uint bpt_balance = balanceOfVaultPool();
```
Comment (batudo) [inline contracts/ControllerBalancerV2Strat.sol:175]:
```suggestion
        uint bptBalance = balanceOfVaultPool();
```

File: contracts/ControllerStratAbs.sol [hunk 3]
```diff
@@ -38,6 +40,13 @@ abstract contract ControllerStratAbs is Swappable, Pausable, ReentrancyGuard {
     address public exchange;
     address public immutable controller; // immutable to prevent anyone to change it and withdraw
 
+    // Deposit compensation
+    address public compensator;
```
Comment (batudo) [inline contracts/ControllerStratAbs.sol:44]:
```suggestion
    address public equalizer;
```
=)

File: contracts/ControllerStratAbs.sol [hunk 4]
```diff
@@ -132,6 +144,22 @@ abstract contract ControllerStratAbs is Swappable, Pausable, ReentrancyGuard {
         rewardToWantRoute[_reward] = _route;
     }
 
+    // Compensation
+    function setCompensateRatio(uint newRatio) external onlyAdmin {
```
Comment (batudo) [inline contracts/ControllerStratAbs.sol:148]:
```suggestion
    function setOffsetRatio(uint newRatio) external onlyAdmin {
```

File: contracts/ControllerStratAbs.sol [hunk 5]
```diff
@@ -38,6 +40,13 @@ abstract contract ControllerStratAbs is Swappable, Pausable, ReentrancyGuard {
     address public exchange;
     address public immutable controller; // immutable to prevent anyone to change it and withdraw
 
+    // Deposit compensation
+    address public compensator;
+    uint public compensateRatio = 0; // 0.00%
```
Comment (batudo) [inline contracts/ControllerStratAbs.sol:45]:
```suggestion
    uint public offsetRatio = 0; // 0.00%
```

File: test/integration/ControllerBalancerV2Strat-test.js [hunk 6]
```diff
@@ -96,13 +98,17 @@ describe('Controller BalancerV2 Strat USDC', () => {
...
+
     // withdraw 95% in shares
     const toWithdraw = (await archimedes.balanceOf(0, bob.address)).mul(
-      8000
+      8500
     ).div(10000)
     let expectedOutput = toWithdraw.mul(await archimedes.getPricePerFullShare(0)).div(1e6)
 
-    await strat.setPoolSlippageRatio(150)
+    // await strat.setPoolSlippageRatio(50)
```
Comment (batudo) [inline test/integration/ControllerBalancerV2Strat-test.js:111]:
```suggestion
```

File: contracts/ControllerStratAbs.sol (hunk 3 above)
Comment (Shelvak) [reply contracts/ControllerStratAbs.sol:44]: fuaaaa Denzel Washington approves :sunglasses:

**Review by batudo — APPROVED** [review]

**Pull request merged**